### DISCUSSION TITLE
Mypy error when using deferred() without Mapped

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#9884] [Usage Questions]
Hello

I have the following mypy error message with SQLAlchemy 2.0.15:

```bash
mypy sample.py --no-strict-optional
sample.py:11: error: Need type annotation for "password"
Found 1 error in 1 file (checked 1 source file)
```

```python
from sqlalchemy import String, Integer
from sqlalchemy.orm import DeclarativeBase, Mapped, deferred, mapped_column

class Base(DeclarativeBase):
    pass

class Users(Base):
    __tablename__ = "users"

    id = mapped_column(Integer, primary_key=True)
    password = deferred(mapped_column("Password", String)) # ko
    password2: Mapped[str] = deferred(mapped_column("Password", String)) # ok
```

Is it a valid use case?
Currently I'm not interested in null safety, so I consider migrating my table models to use `mapped_column` without explicit type annotations.

**Comment by nmoreaud** [comment]
Note: example taken from issue https://github.com/sqlalchemy/sqlalchemy/issues/9536 but without `Mapped` type annotation

**Comment by CaselIT** [comment]
Hi,

That's expected. There is really nothing sqlalchemy can do to change that error.

There may be a mypy setting to suppress that error.

**Comment by nmoreaud** [comment]
The right way to do is to replace
`password = deferred(mapped_column("Password", String))`
with
`password = mapped_column("Password", String, deferred=True)`
?

Is it semantically the same?

**Reply by CaselIT** [reply]
It is, but if you plan to use mypy you should really look into using `Mapped[]`

**Reply by zzzeek** [reply]
if mypy can implicitly recognize other attributes then it's reasonable for someone to want that.


to the poster: just use mapped_column(deferred=True).  this is identical to using deferred()

**Reply by CaselIT** [reply]
deferred seems to be already correctly typed:
```py
def deferred(column: _ORMColumnExprArgument[_T], ...)->MappedSQLExpression[_T]:
```

I guess the issue is that `mapped_column` returns `MappedColumn[Any]` so mypy does not know what to set to `_T` in deferred?

That said I'm not really sure we can do anything to fix this, since we want that `Any` as the return type of mapped_column

**Reply by zzzeek** [reply]
mapped_column() should be returning MappedColumn[_T] if there's a TypeEngine argument explicitly present

**Reply by CaselIT** [reply]
I think we decided for a no in that case, since it would prevent people from explicitly typing the left side with something different
